### PR TITLE
[WIP]: Config read initial attempt

### PR DESCRIPTION
**Pull request by crutchcorn**
This is an initial WIP PR to migrate the [config reading logic from `isomorphic-git`](https://github.com/isomorphic-git/isomorphic-git/blob/main/src/models/GitConfig.js#L32-L151) to GitOxide

There's still work need to do to read and write values from this parsed config (and to normalize the data storage).

I'm going to be requesting a code review from a colleague as this is the first time I've written Rust code.

I've also added initially functional tests (although we will need to modify the tests as we improve and evolve the API)

### CODE REVIEW

File: git-config/src/parse.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    // variable lines contain a name, and equal sign and then a value
+    // variable lines can also only contain a name (the implicit value is a boolean true)
+    // variable name is alphanumeric (ASCII) with -
+    // variable name starts with an alphabetic character
+    // variable name is case insensitive
+    static ref VARIABLE_LINE_REGEX: Result<regex::Regex, regex::Error> = Regex::new(r#"^([A-Za-z][A-Za-z-]*)(?: *= *(.*))?$"#);
+    static ref VARIABLE_NAME_REGEX: Result<regex::Regex, regex::Error> = Regex::new(r#"^[A-Za-z][A-Za-z-]*$"#);
+
+    static ref VARIABLE_VALUE_COMMENT_REGEX: Result<regex::Regex, regex::Error> = Regex::new(r#"^(.*?)( *[#;].*)$"#);
+}
```
Comment (crutchcorn) [inline git-config/src/parse.rs:25]: I would have liked to avoided using `lazy_static!`, but I was getting errors without doing this trick.

I also would have liked to `unwrap()` each of these `Regex::new` calls, but because `Debug` is not present at this top level environment, the compiler disallowed me from doing so. As such, I have to `unwrap` every instance or running these regexes

File: git-config/src/parse.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    matches.and_then(|cap| {
+        match (cap.get(1), cap.get(2)) {
+            (Some(cap1), Some(cap2)) => Some((cap1.as_str(), cap2.as_str())),
+            _ => { None }
+        }
+    })
+}
+
+fn extract_variable_line(line: &str) -> Option<(String, String)> {
+    let matches = VARIABLE_LINE_REGEX.as_ref().unwrap().captures(line);
```
Comment (crutchcorn) [inline git-config/src/parse.rs:39]: Notice the `as_ref` side effect from having to use `lazy_static` and the `unwrap` from not having `Debug` present :/

File: git-config/src/parse.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    return matches.map(|cap|
+        (
+            cap.get(1).map(|m| m.as_str()),
+            cap.get(2).map(|m| m.as_str()).unwrap_or("true")
+        )
+    )
+        .and_then(|(name_opt, raw_value_opt)| {
```
Comment (crutchcorn) [inline git-config/src/parse.rs:47]: I've learned that `and_then` is sort of a `flatmap` on `Option`

> Returns None if the option is None, otherwise calls f with the wrapped value and returns the result.
>
> Some languages call this operation flatmap.

https://doc.rust-lang.org/beta/std/option/enum.Option.html#method.and_then

File: git-config/src/parse.rs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            match (name_opt, raw_value_opt) {
+                (Some(name), raw_value) => Some((name, raw_value)),
+                _ => { None }
+            }
+        })
+        .map(|(name, raw_value)| {
+            let value_without_comments = remove_comments(raw_value);
+            let value_without_quotes = remove_quotes(&value_without_comments);
```
Comment (crutchcorn) [inline git-config/src/parse.rs:55]: I was getting compiler errors if I didn't use `&String`

File: git-config/src/parse.rs [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        }) {
+        Some(val) => val,
+        _ => { raw_value.to_string() }
+    };
+}
+
+fn has_odd_number_of_quotes(text: &str) -> bool {
+    let quote_regex = Regex::new(r#"(?g)(?:^|[^\\])""#).unwrap();
+    let number_of_quotes = quote_regex.captures(text);
+    return number_of_quotes.unwrap().len() % 2 != 0;
```
Comment (crutchcorn) [inline git-config/src/parse.rs:91]: I may refactor this to not use `unwrap` - I would like to avoid using `unwrap` whenever possible, as I know it's a failure point in the system and can crash the program

File: git-config/src/parse.rs (hunk 1 above)
Comment (d3lm) [reply git-config/src/parse.rs:25]: Use `once_cell` instead and `Lazy`. Much more convenient to work with IMO and allows to lazily create and unwrap the regex. Created a small playground for you: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=baf7bb46640b30d24db0808251e27a98

File: git-config/src/parse.rs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    return matches.map(|cap|
+        (
+            cap.get(1).map(|m| m.as_str()),
+            cap.get(2).map(|m| m.as_str()).unwrap_or("true")
+        )
+    )
+        .and_then(|(name_opt, raw_value_opt)| {
+            match (name_opt, raw_value_opt) {
+                (Some(name), raw_value) => Some((name, raw_value)),
+                _ => { None }
```
Comment (d3lm) [inline git-config/src/parse.rs:50]: FYI, there is no need for the curly braces.

File: git-config/src/parse.rs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            let value_without_quotes = remove_quotes(&value_without_comments);
+            (name.to_string(), value_without_quotes)
+        });
+}
+
+// removeComments
+fn remove_comments(raw_value: &str) -> String {
+    let comment_matches = VARIABLE_VALUE_COMMENT_REGEX.as_ref().unwrap().captures(raw_value);
+
+    return match comment_matches
```
Comment (d3lm) [inline git-config/src/parse.rs:64]: I would recommend splitting this function chain and the match statement apart. Really hard to read. Do the transformation first, put it into a variable, and then match it.

File: git-config/src/parse.rs (hunk 4 above)
Comment (d3lm) [reply git-config/src/parse.rs:55]: There are differences between `String`, `&String` or `&str` or even `str`, which is the section in memory for the string. 

A `&str` is just a pointer + length. It's a slice, kind of like `Vec<X>`
A `String` is a pointer + length + capacity, and is owned.

So `String` is a struct, and if you do `&String` it's a reference to that struct which has a pointer, length and capacity. In other words, `&String` is not different to a ref to any struct in Rust. However, it's very rarely more useful than `&str`, in fact, the latter is preferred in most cases because it's more generic. `&String` is a double indirection, cause `String` is already one level of indirection. It's a pointer to a struct that contains a pointer. 2 levels of indirection for `&String`. Keep those things in mind when you write Rust.

Bottom line: always prefer `&str` over `&String`

File: git-config/src/parse.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    };
+}
+
+fn has_odd_number_of_quotes(text: &str) -> bool {
+    let quote_regex = Regex::new(r#"(?g)(?:^|[^\\])""#).unwrap();
+    let number_of_quotes = quote_regex.captures(text);
+    return number_of_quotes.unwrap().len() % 2 != 0;
+}
+
+fn remove_quotes(text: &String) -> String {
```
Comment (d3lm) [inline git-config/src/parse.rs:94]: This function should accept `&str` and not `&String`.

File: git-config/src/parse.rs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    let mut new_text = "".to_owned();
```
Comment (d3lm) [inline git-config/src/parse.rs:95]: `to_owned()` actually allocates memory, whereas `String::new()` doesn't. Also, the compiler won't optimize your code into something more efficient such as `String::new()`. For empty owned String's always prefer the latter.

File: git-config/src/parse.rs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        let is_escape_for_quote = c == "\\" && text.chars().nth(idx + 1).unwrap_or(' ') == '"';
+        if !is_quote && !is_escape_for_quote {
+            new_text.push_str(c);
+        }
+    }
+
+    return new_text;
+}
+
+fn get_path(section: &str, subsection: &str, name: &String) -> String {
```
Comment (d3lm) [inline git-config/src/parse.rs:107]: Same here, use `&str` for all parameters.

File: git-config/src/parse.rs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    pub subsection: String,
+    pub name: String,
+    pub value: String,
+    pub path: String
+}
+
+// Note: there are a LOT of edge cases that aren't covered (e.g. keys in sections that also
+// have subsections, [include] directives, etc.
+pub fn parse_config(text: String) -> Vec<ParsedConfig> {
+    let mut section: &str = "";
```
Comment (d3lm) [inline git-config/src/parse.rs:125]: Type annotations are not necessarily needed here. Just saying.

File: git-config/src/parse.rs [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    let mut subsection: &str = "";
+
+    let parsed_config = text.split("\n").map(|line| {
+        let mut name: String = "".to_string();
```
Comment (d3lm) [inline git-config/src/parse.rs:129]: Use `String::new()`.

File: git-config/src/parse.rs [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        let mut value: String = "".to_string();
```
Comment (d3lm) [inline git-config/src/parse.rs:130]: Same

File: git-config/tests/read/mod.rs [hunk 14]
```diff
@@ -0,0 +1,27 @@
+mod read {
+    use git_config::parse::parse_config;
+    use git_config::ParsedConfig;
+
+    #[test]
+    fn read_test_file() -> crate::Result {
+        const CONFIG: &str = include_str!("../assets/config");
+
+        assert_eq!(parse_config(CONFIG.to_string()), vec![
+            ParsedConfig { is_section: false, section: "".to_string(), subsection: "".to_string(), name: "".to_string(), value: "".to_string(), path: "".to_string() },
```
Comment (d3lm) [inline git-config/tests/read/mod.rs:10]: I see a lot of `"".to_string()`, replace them with `String::new()`.

**Review by d3lm — COMMENTED** [review]
Left a couple of remarks. Other than that, looking good 👍

File: git-config/src/parse.rs (hunk 4 above)
Comment (crutchcorn) [reply git-config/src/parse.rs:55]: Gotcha. That's what I originally thought as well, but truthfully I ran into compiler errors when using `&str` and just kinda guess stopped them. Bad engineering, for sure - I'll go back and try to revert back to `&str` and ask you specifically how to _properly_ fix the problems

File: git-config/src/parse.rs (hunk 7 above)
Comment (crutchcorn) [reply git-config/src/parse.rs:64]: Makes sense - I'll do just that

File: git-config/src/parse.rs (hunk 9 above)
Comment (crutchcorn) [reply git-config/src/parse.rs:95]: Admittedly, this was a copy+paste job from StackOverflow to get `push_str` working. I thought `owned` might have been a different data type of some kind or something. I'll switch that

File: git-config/src/parse.rs (hunk 12 above)
Comment (crutchcorn) [reply git-config/src/parse.rs:129]: Just curious: What's the difference under-the-hood between `to_string` and `String::new`? Is it similar to the memory allocation thing you were talking about with `.to_owned`?

**Review by crutchcorn — COMMENTED** [review]
Awesome! Thanks a bunch for the review @d3lm, it was super helpful!

File: git-config/tests/read/mod.rs (hunk 14 above)
Comment (d3lm) [reply git-config/tests/read/mod.rs:10]: Well, the truth is, under the hood `to_string` also doesn't allocate but it's just that before Rust 1.9.0 it was slower than `String::new()` and it uses the `Display` trait. So it was doing `String::new()` and then called `.write_fmt()` on in general does more work. So just calling `String::new()` is faster, however that is not entirely true anymore after 1.9.0. But the latter is still more idiomatic.
